Detection of cpuid producing false positive

When building on `ppc64el` architecture, the `cpuid` detection in [dfxml_configure.m4](https://github.com/dfxml-working-group/dfxml_cpp/blob/09030bc5fb127c84cae6552f2e37961bc93c51ac/src/dfxml_configure.m4#L36) appears to be producing a false positive.

I added a debug statement and found that `have_cpuid` is set to "yes".

```
[...]
checking for git... no
configure: WARNING: git not found
configure: have_cpuid: yes
checking for CommonCrypto/CommonDigest.h... no
checking for gcrypt.h... no
[...]
```

This results in an assembler error.

```
be20_api/dfxml_cpp/src/dfxml_writer.h:478:96: warning: macro "__TIME__" might prevent reproducible builds [-Wdate-time]
  478 |             snprintf(buf,sizeof(buf),"%4d-%02d-%02dT%s",tm.tm_year+1900,tm.tm_mon+1,tm.tm_mday,__TIME__);
      |                                                                                                ^~~~~~~~
/tmp/cc8zqd7t.s: Assembler messages:
/tmp/cc8zqd7t.s:47103: Error: unrecognized opcode: `cpuid'
/tmp/cc8zqd7t.s:48665: Error: unrecognized opcode: `cpuid'
/tmp/cc8zqd7t.s:55129: Error: unrecognized opcode: `cpuid'
make[3]: *** [Makefile:1383: phase1.o] Error 1
```

Neat. I do not have access to a system to test. Can you add a #ifdef to turn it off on this platform?


----
Sent from my phone.

On May 19, 2022, at 11:19 PM, Jason ***@***.***> wrote:

﻿
When building on ppc64el architecture, the cpuid detection in dfxml_configure.m4 appears to be producing a false positive.

I added a debug statement and found that have_cpuid is set to "yes".

[...]
checking for git... no
configure: WARNING: git not found
configure: have_cpuid: yes
checking for CommonCrypto/CommonDigest.h... no
checking for gcrypt.h... no
[...]
This results in an assembler error.

be20_api/dfxml_cpp/src/dfxml_writer.h:478:96: warning: macro "__TIME__" might prevent reproducible builds [-Wdate-time]
  478 |             snprintf(buf,sizeof(buf),"%4d-%02d-%02dT%s",tm.tm_year+1900,tm.tm_mon+1,tm.tm_mday,__TIME__);
      |                                                                                                ^~~~~~~~
/tmp/cc8zqd7t.s: Assembler messages:
/tmp/cc8zqd7t.s:47103: Error: unrecognized opcode: `cpuid'
/tmp/cc8zqd7t.s:48665: Error: unrecognized opcode: `cpuid'
/tmp/cc8zqd7t.s:55129: Error: unrecognized opcode: `cpuid'
make[3]: *** [Makefile:1383: phase1.o] Error 1
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you are subscribed to this thread.

--- a/src/dfxml_configure.m4
+++ b/src/dfxml_configure.m4
@@ -33,7 +33,19 @@ AM_COND_IF([FOUND_GIT],
         [AC_MSG_WARN([git not found])])
 
 # Do we have the CPUID instruction?
-AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#define cpuid(id) __asm__( "cpuid" : "=a"(eax), "=b"(ebx), "=c"(ecx), "=d"(edx) : "a"(id), "b"(0), "c"(0), "d"(0))]], [[unsigned long eax, ebx, ecx, edx;cpuid(0);]])],[have_cpuid=yes],[have_cpuid=no])
+# Based on https://www.gnu.org/software/autoconf-archive/ax_gcc_x86_cpuid.html
+AC_RUN_IFELSE([AC_LANG_PROGRAM([int eax, ebx, ecx, edx;], [
+    __asm__ __volatile__ ("xchg %%ebx, %1\n"
+      "cpuid\n"
+      "xchg %%ebx, %1\n"
+      : "=a" (eax), "=r" (ebx), "=c" (ecx), "=d" (edx)
+      : "a" (0), "2" (0));
+    return 0;
+])],
+[have_cpuid=yes],
+[have_cpuid=no],
+[have_cpuid=no])
+AC_MSG_NOTICE([have_cpuid: $have_cpuid])
 if test "$have_cpuid" = yes; then
   AC_DEFINE(HAVE_ASM_CPUID, 1, [define to 1 if __asm__ CPUID is available])
 fi

